### PR TITLE
chore(input): refactor keyboard layout, extract pure layout

### PR DESCRIPTION
**Pull request by pavelfeldman**

### CODE REVIEW

File: src/input.ts [hunk 1]
```diff
@@ -246,3 +216,64 @@ export class Mouse {
     await this.click(x, y, { ...options, clickCount: 2 });
   }
 }
+
+const aliases = new Map<string, string[]>([
+  ['ShiftLeft', ['Shift']],
```
Comment (JoelEinbinder) [inline src/input.ts:221]: These should come from when key !== code. I'd rather have some logic about the ordering in the key definitions map then the aliases here.

File: src/input.ts [hunk 2]
```diff
@@ -246,3 +216,64 @@ export class Mouse {
...
+  ['ControlLeft', ['Control']],
+  ['AltLeft', ['Alt']],
+  ['MetaLeft', ['Meta']],
+  ['Enter', ['\n', '\r']],
```
Comment (JoelEinbinder) [inline src/input.ts:225]: Let's do the same with `description.text` as `description.key`, and remove the `\r` here. Then we are left with just one line adding `\n`.

File: src/input.ts [hunk 3]
```diff
@@ -246,3 +216,64 @@ export class Mouse {
...
+    result.set(code, { ...description, shifted: shiftedDescription });
+
+    // Map from aliases: Shift -> non-shiftable definition
+    if (aliases.has(code)) {
+      for (const alias of aliases.get(code)!)
+        result.set(alias, description);
+    }
+
+    // Do not use numpad when converting keys to codes.
+    if (definition.location)
```
Comment (JoelEinbinder) [inline src/input.ts:267]: This can be removed if order matters in usKeyboardLayout

File: src/input.ts [hunk 4]
```diff
@@ -246,3 +216,64 @@ export class Mouse {
...
+      text: definition.text || '',
+      location: definition.location || 0,
+    };
+    if (definition.key.length === 1)
+      description.text = description.key;
+
+    // Generate shifted definition.
+    let shiftedDescription: KeyDescription | undefined;
+    if (definition.shiftKey) {
+      assert(definition.shiftKey.length === 1);
```
Comment (JoelEinbinder) [inline src/input.ts:248]: I believe this assertion is coincidental.

File: src/input.ts [hunk 5]
```diff
@@ -246,3 +216,64 @@ export class Mouse {
...
+      shiftedDescription = { ...description };
+      shiftedDescription.key = definition.shiftKey;
+      shiftedDescription.text = definition.shiftKey;
+      if (definition.shiftKeyCode)
+        shiftedDescription.keyCode = definition.shiftKeyCode;
+      description.shifted = shiftedDescription;
+    }
+
+    // Map from code: Digit3 -> { ... descrption, shifted }
+    result.set(code, { ...description, shifted: shiftedDescription });
```
Comment (JoelEinbinder) [inline src/input.ts:258]: `description.shifted` is already set, so this is equivalent to result.set(code, description);
```suggestion
    result.set(code, description);
```

`shiftedDescription` can be const inside the if block as well then.

**Review by JoelEinbinder — APPROVED** [review]
I liked this refactor a lot! I want my fancy keys back though.

File: src/macEditingCommands.ts [hunk 6]
```diff
@@ -0,0 +1,133 @@
+/**
+ * Copyright 2017 Google Inc. All rights reserved.
```
[the remark this answers is not in the document]
Comment (pavelfeldman) [reply src/macEditingCommands.ts:2]: To be on a safe side, when I extract portion of the file into its own file, I take it's header.

File: src/usKeyboardLayout.ts [hunk 7]
```diff
@@ -15,392 +15,140 @@
...
-  '"': {'keyCode': 222, 'key': '"', 'code': 'Quote'},
-  'SoftLeft': {'key': 'SoftLeft', 'code': 'SoftLeft', 'location': 4},
-  'SoftRight': {'key': 'SoftRight', 'code': 'SoftRight', 'location': 4},
-  'Camera': {'keyCode': 44, 'key': 'Camera', 'code': 'Camera', 'location': 4},
-  'Call': {'key': 'Call', 'code': 'Call', 'location': 4},
-  'EndCall': {'keyCode': 95, 'key': 'EndCall', 'code': 'EndCall', 'location': 4},
-  'VolumeDown': {'keyCode': 182, 'key': 'VolumeDown', 'code': 'VolumeDown', 'location': 4},
-  'VolumeUp': {'keyCode': 183, 'key': 'VolumeUp', 'code': 'VolumeUp', 'location': 4},
-};
+export const Qwerty: KeyboardLayout = {
```
[the remark this answers is not in the document]
Comment (pavelfeldman) [reply src/usKeyboardLayout.ts:32]: Renamed back to USKeyboardLayout.

File: src/input.ts [hunk 8]
```diff
@@ -144,10 +108,16 @@ export class Keyboard {
   }
 
   async press(key: string, options: { delay?: number } = {}) {
+    const tokens = key.split('+');
```
[the remark this answers is not in the document]
Comment (pavelfeldman) [reply src/input.ts:111]: Done

File: src/input.ts (hunk 1 above)
Comment (pavelfeldman) [reply src/input.ts:221]: Could you elaborate?

File: src/input.ts (hunk 2 above)
Comment (pavelfeldman) [reply src/input.ts:225]: I don't get this one either :/

File: src/input.ts (hunk 3 above)
Comment (pavelfeldman) [reply src/input.ts:267]: Yeah, but it becomes subtle that way. I'd rather use a bigger hammer.

File: src/input.ts (hunk 4 above)
Comment (pavelfeldman) [reply src/input.ts:248]: I actually make sure shiftKey is short at all times - that saves me a check further along.

File: src/input.ts (hunk 5 above)
Comment (pavelfeldman) [reply src/input.ts:258]: It is actually backwards - i want no 'shifted' inside `3` or `#`, only inside `Digit3`.

**Pull request merged**